[SPARK-19413][SS] MapGroupsWithState for arbitrary stateful operations

tdas opened the pull request:
## What changes were proposed in this pull request?

`mapGroupsWithState` is a new API for arbitrary stateful operations in Structured Streaming, similar to `DStream.mapWithState` 

*Requirements*
- Users should be able to specify a function that can do the following
- Access the input row corresponding to a key
- Access the previous state corresponding to a key
- Optionally, update or remove the state
- Output any number of new rows (or none at all)

*Proposed API*
```
// ------------ New methods on KeyValueGroupedDataset ------------
class KeyValueGroupedDataset[K, V] {	
	// Scala friendly
	def mapGroupsWithState[S: Encoder, U: Encoder](func: (K, Iterator[V], KeyedState[S]) => U)
        def flatMapGroupsWithState[S: Encode, U: Encoder](func: (K, Iterator[V], KeyedState[S]) => Iterator[U])
	// Java friendly
       def mapGroupsWithState[S, U](func: MapGroupsWithStateFunction[K, V, S, R], stateEncoder: Encoder[S], resultEncoder: Encoder[U])
       def flatMapGroupsWithState[S, U](func: FlatMapGroupsWithStateFunction[K, V, S, R], stateEncoder: Encoder[S], resultEncoder: Encoder[U])
}

// ------------------- New Java-friendly function classes ------------------- 
public interface MapGroupsWithStateFunction<K, V, S, R> extends Serializable {
  R call(K key, Iterator<V> values, state: KeyedState<S>) throws Exception;
}
public interface FlatMapGroupsWithStateFunction<K, V, S, R> extends Serializable {
  Iterator<R> call(K key, Iterator<V> values, state: KeyedState<S>) throws Exception;
}

// ---------------------- Wrapper class for state data ---------------------- 
trait KeyedState[S] {
	def exists(): Boolean  	
  	def get(): S 			// throws Exception is state does not exist
	def getOption(): Option[S]       
	def update(newState: S): Unit
	def remove(): Unit		// exists() will be false after this
}
```

Key Semantics of the State class
- The state can be null.
- If the state.remove() is called, then state.exists() will return false, and getOption will returm None.
- After that state.update(newState) is called, then state.exists() will return true, and getOption will return Some(...). 
- None of the operations are thread-safe. This is to avoid memory barriers.

*Usage*
```
val stateFunc = (word: String, words: Iterator[String, runningCount: KeyedState[Long]) => {
    val newCount = words.size + runningCount.getOption.getOrElse(0L)
    runningCount.update(newCount)
   (word, newCount)
}

dataset					                        // type is Dataset[String]
  .groupByKey[String](w => w)        	                // generates KeyValueGroupedDataset[String, String]
  .mapGroupsWithState[Long, (String, Long)](stateFunc)	// returns Dataset[(String, Long)]
```


## How was this patch tested?
New unit tests.

tdas committed:
Prototype - almost working

tdas committed:
Renamed to mapGroupsWithState

tdas committed:
Fixed bugs

tdas committed:
Removed prints

tdas committed:
Test state remove

tdas committed:
Test restart, and test with metrics

tdas committed:
Fixed everything

tdas committed:
Refactored, added java APIs and tests

tdas committed:
Refactored

tdas committed:
Added more test

tdas committed:
Merge remote-tracking branch 'apache-github/master' into mapWithState

tdas committed:
Added docs

SparkQA commented:
**[Test build #72205 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72205/testReport)** for PR 16758 at commit [`6fab7a5`](https://github.com/apache/spark/commit/6fab7a5fde75309198d1e73e66948d82d0f590e6).

SparkQA commented:
**[Test build #72205 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72205/testReport)** for PR 16758 at commit [`6fab7a5`](https://github.com/apache/spark/commit/6fab7a5fde75309198d1e73e66948d82d0f590e6).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72205/
Test FAILed.

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulAggregate.scala:
```diff
@@ -54,6 +55,18 @@ trait StatefulOperator extends SparkPlan {
   }
 }
 
+trait StateStoreReader extends StatefulOperator {
```
This files should probably be renamed from StatefulAggregation to StatefulOperations.

marmbrus replied:
and lowercase if it contains multiple classes

tdas committed:
Added docs

SparkQA commented:
**[Test build #72206 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72206/testReport)** for PR 16758 at commit [`8be63de`](https://github.com/apache/spark/commit/8be63de507f9e4fc258aca1d467509e06b674565).

SparkQA commented:
**[Test build #72206 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72206/testReport)** for PR 16758 at commit [`8be63de`](https://github.com/apache/spark/commit/8be63de507f9e4fc258aca1d467509e06b674565).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72206/
Test FAILed.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala:
```diff
@@ -219,6 +219,160 @@ class KeyValueGroupedDataset[K, V] private[sql](
+   * [[Dataset]].
+   *
+   * This function can be applied on both batch and streaming Datasets. With a streaming dataset,
+   * this function will be once for each in every trigger. For each key, the updated state from the
```
will be *called*

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala:
```diff
@@ -219,6 +219,160 @@ class KeyValueGroupedDataset[K, V] private[sql](
+   *
+   * This function can be applied on both batch and streaming Datasets. With a streaming dataset,
+   * this function will be once for each in every trigger. For each key, the updated state from the
+   * function call in a trigger will be the state available in the function call in the next
```
Any updates to the state will be stored and passed to the user given function in subsequent batches when executed as a Streaming Query.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala:
```diff
@@ -219,6 +219,160 @@ class KeyValueGroupedDataset[K, V] private[sql](
   /**
+   * ::Experimental::
+   * (Scala-specific)
+   * Applies the given function to each group of data, while using an additional keyed state.
```
while maintaining some user-defined state for each key.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala:
```diff
@@ -219,6 +219,160 @@ class KeyValueGroupedDataset[K, V] private[sql](
+   * ::Experimental::
+   * (Scala-specific)
+   * Applies the given function to each group of data, while using an additional keyed state.
+   * For each unique group, the function will be passed the group key and an iterator that contains
```
I would consider breaking this up to make it a little easier to follow:

```
For each unique group, the given function will be invoked with the following arguments:
 - The key of the group.
 - A user-defined state object set by previous invocations of the given function.  Note that, for batch queries, there is only ever one invocation and thus the state object will always be empty.
 - An iterator containing all the values for this key.
```

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala:
```diff
@@ -219,6 +219,160 @@ class KeyValueGroupedDataset[K, V] private[sql](
+   * trigger. However, for batch, `mapGroupsWithState` behaves exactly as `mapGroups` and the
+   * function is called only once per key without any prior state.
+   *
+   * There is no guaranteed ordering of values in the iterator in the function.
```
I'd maybe put these into bullets as well.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala:
```diff
@@ -219,6 +219,160 @@ class KeyValueGroupedDataset[K, V] private[sql](
+   */
+  @Experimental
+  @InterfaceStability.Evolving
+  def mapGroupsWithState[STATE: Encoder, OUT: Encoder](
```
I think it would be useful to come up with an example use case since this is pretty complicated.

marmbrus replied:
I would also stick with `S` and `U` to match other functions in the class.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala:
```diff
@@ -219,6 +219,160 @@ class KeyValueGroupedDataset[K, V] private[sql](
+   */
+  @Experimental
+  @InterfaceStability.Evolving
+  def mapGroupsWithState[STATE, OUT](
```
Since this is pretty complicated and we might iterate over time, I wonder if we shouldn't just put the full explanation in `State` and link there with a very brief version for each function.  Not sure...

tdas replied:
Right now I have the detailed example in `State`, but all the main points to note are in the docs for `mapGroupsWithState`. That's kind of inconsistent. 

There are 5 places where there should be docs - 2 mapGroupsWithState, 2 flatMapGroupsWithState and KeyedState. How about we do the following.
- Scala-friendly versions of mapGroupsWithState and flatMapGRoupsWithState have the full explanation with examples.
- Java-friendly APIspoint to the scala-friendly ones, 
- KeyedState points to the mapGroupsWithState and flatMapGRoupsWithState.

This will reduce it to just 2 authoritative locations.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/State.scala:
```diff
@@ -0,0 +1,101 @@
+/**
+ * :: Experimental ::
+ *
+ * Wrapper class for interacting with state data in `mapGroupsWithState` and
```
I think we want to say what state we are talking about.  Something like "per-key state from previous invocations of the function in a StreamingQuery"

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/State.scala:
```diff
@@ -0,0 +1,101 @@
+ */
+@Experimental
+@InterfaceStability.Evolving
+trait State[S] extends LogicalState[S] {
```
`KeyState`?  State just feels very generic.

tdas replied:
Agreed.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/State.scala:
```diff
@@ -0,0 +1,101 @@
+ *       };
+ * }}}
+ *
+ * @tparam S Type of the state
```
A user defined type that can be stored for each key.  Must be encodable?

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/State.scala:
```diff
@@ -0,0 +1,101 @@
+
+  def exists: Boolean
+
+  def get(): S
```
No `()`.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/State.scala:
```diff
@@ -0,0 +1,101 @@
+
+  def remove(): Unit
+
+  @inline final def getOption(): Option[S] = if (exists) Some(get()) else None
```
No `()`.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/State.scala:
```diff
@@ -0,0 +1,101 @@
+
+  def update(newState: S): Unit
+
+  def remove(): Unit
```
Scala doc for these, even though its pretty obvious.  In particular, I assume its safe to call update() get remove() more than once in the function?

tdas replied:
Will do.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:
```diff
@@ -313,6 +313,25 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
+  object MapGroupsWithStateStrategy extends Strategy {
+    override def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
+      case MapGroupsWithState(
+        func, keyDeser, valueDeser, groupAttr, dataAttr, outputAttr,
```
I think we indent 4 or 2 past the class normally.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/StateImpl.scala:
```diff
@@ -0,0 +1,70 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
```
Should this be in `streaming`?

tdas replied:
I had so in the prototype, but then State is used in both batch and streaming. Hence I shifted it from `execution.streaming` to `streaming`.

marmbrus replied:
Yeah, I'd probably still go with streaming since thats its primary use.

tdas replied:
This is internal so thats fine. However, should KeyedState trait be in the `o.a.s.sql` or `o.a.s.sql.streaming`?

marmbrus replied:
Good question.   I would go with streaming to keep the top level clean.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/StateImpl.scala:
```diff
@@ -0,0 +1,70 @@
+  private var value: S = optionalValue.getOrElse(null.asInstanceOf[S])
+  private var defined: Boolean = optionalValue.isDefined
+  private var updated: Boolean = false  // whether value has been updated (but not removed)
+  private var removed: Boolean = false  // whether value has eben removed
```
"been"

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/StateImpl.scala:
```diff
@@ -0,0 +1,70 @@
+import org.apache.spark.sql.State
+
+/** Internal implementation of the [[State]] interface */
+private[sql] class StateImpl[S](optionalValue: Option[S]) extends State[S] {
```
I would consider using `null` here to avoid extra allocations in the critical path.

tdas replied:
I believe null can be a valid state value, so just taking null does not disambiguate whether state exists and is null v/s state does not exist. You would have to do `class StateImpl[S](value: S, isDefined: Boolean)`.  But this does not avoid object wrapping because, the Option wrapping is currently done by the StateStore implementation. We will have to change StateStore interfaces to avoid that. So I think its premature to do this in this PR.

marmbrus replied:
Wait, I do not think that we should ever return `Some(null)`.  In general I would expect `null` to mean there is no state, similar to gets in a java hashtable.

tdas replied:
So will `KeyedState.update(null)` be equivalent to `remove()`?

marmbrus replied:
Yes, I think that makes sense (since a `get()` should later return `null`).

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/StateImpl.scala:
```diff
@@ -0,0 +1,70 @@
+    defined
+  }
+
+  override def get(): S = {
```
no `()`

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala:
```diff
@@ -90,6 +93,14 @@ class IncrementalExecution(
               Some(stateId),
               child) :: Nil))
+      case MapGroupsWithStateExec(
+        func, keyDeser, valueDeser, groupAttr, dataAttr, outputAttr,
```
indent inconsistent with above.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulAggregate.scala:
```diff
@@ -54,6 +55,18 @@ trait StatefulOperator extends SparkPlan {
+    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"))
+}
+
+trait StateStoreWriter extends StatefulOperator {
```
Scala doc for both of these.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulAggregate.scala:
```diff
@@ -235,3 +240,86 @@ case class StateStoreSaveExec(
+
+  override protected def doExecute(): RDD[InternalRow] = {
+
+    child.execute().mapPartitionsWithStateStore[InternalRow](
```
I don't think you need the type here?

tdas replied:
no harm keeping it. the function below is a little complex and it throws better compiler error message while developing if you goof up the function returns.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulAggregate.scala:
```diff
@@ -235,3 +240,86 @@ case class StateStoreSaveExec(
+    ClusteredDistribution(groupingAttributes) :: Nil
+
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] =
+    Seq(groupingAttributes.map(SortOrder(_, Ascending)))   // is this ordering needed?
```
Yes, the GroupedIterator relies on sorting.

tdas replied:
Thanks for confirming, will remove the comment.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulAggregate.scala:
```diff
@@ -235,3 +240,86 @@ case class StateStoreSaveExec(
+    Seq(groupingAttributes.map(SortOrder(_, Ascending)))   // is this ordering needed?
+
+  override protected def doExecute(): RDD[InternalRow] = {
+
```
nit: extra newline

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulAggregate.scala:
```diff
@@ -235,3 +240,86 @@ case class StateStoreSaveExec(
+      child.output.toStructType,
+      sqlContext.sessionState,
+      Some(sqlContext.streams.stateStoreCoordinator)) { (store, iter) =>
+        try {
```
existing, but should `mapPartitionsWithStateStore` be implementing the abort handling.  Seems generic.

tdas replied:
I thought of that. If I change the mapPartitionsWithStateStore, then StateStoreSave should also need to be change, which I didnt want to touch in this already big PR

tdas replied:
Managed to do it. Not a big change. Improves correctness, and reduces code.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulAggregate.scala:
```diff
@@ -184,7 +189,7 @@ case class StateStoreSaveExec(
 
             // Assumption: Append mode can be done only when watermark has been specified
-            store.remove(watermarkPredicate.get.eval)
+            store.remove(watermarkPredicate.get.eval _)
```
Why this change?

tdas replied:
Because of the addition of a new version of `stateStore.remove`. I think the compiler can disambiguate correctly without this.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala:
```diff
@@ -144,6 +145,12 @@ object ObjectOperator {
   }
 
+  def deserializeRowToObject(
+    deserializer: Expression): InternalRow => Any = {
```
indent, also does this not fit?

marmbrus reviewed (commented):
Looking pretty good!

tdas committed:
Addressed many comments

SparkQA commented:
**[Test build #72215 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72215/testReport)** for PR 16758 at commit [`34449e4`](https://github.com/apache/spark/commit/34449e4c9de2ed8bd3b0d81ec7bc31580b066420).

tdas committed:
Addressed more comments

SparkQA commented:
**[Test build #72215 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72215/testReport)** for PR 16758 at commit [`34449e4`](https://github.com/apache/spark/commit/34449e4c9de2ed8bd3b0d81ec7bc31580b066420).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72215/
Test FAILed.

SparkQA commented:
**[Test build #72216 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72216/testReport)** for PR 16758 at commit [`3628af8`](https://github.com/apache/spark/commit/3628af82fb85e2a6dc145af5b60dfb8d23f71819).

SparkQA commented:
**[Test build #72216 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72216/testReport)** for PR 16758 at commit [`3628af8`](https://github.com/apache/spark/commit/3628af82fb85e2a6dc145af5b60dfb8d23f71819).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72216/
Test FAILed.

tdas committed:
Addresed more comments

tdas committed:
Added missing classes

SparkQA commented:
**[Test build #72217 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72217/testReport)** for PR 16758 at commit [`59c229b`](https://github.com/apache/spark/commit/59c229b0be934b643950e40ac75051f22b756c93).

tdas committed:
Addressed 1 comment

tdas committed:
Merge remote-tracking branch 'apache-github/master' into mapWithState

SparkQA commented:
**[Test build #72223 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72223/testReport)** for PR 16758 at commit [`3dc0353`](https://github.com/apache/spark/commit/3dc035369be942ffa4c0630c3a23ea8e32a107e1).

SparkQA commented:
**[Test build #72217 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72217/testReport)** for PR 16758 at commit [`59c229b`](https://github.com/apache/spark/commit/59c229b0be934b643950e40ac75051f22b756c93).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72217/
Test FAILed.

SparkQA commented:
**[Test build #72223 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72223/testReport)** for PR 16758 at commit [`3dc0353`](https://github.com/apache/spark/commit/3dc035369be942ffa4c0630c3a23ea8e32a107e1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72223/
Test FAILed.

tdas committed:
Addressed comments

SparkQA commented:
**[Test build #72230 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72230/testReport)** for PR 16758 at commit [`8b18fa1`](https://github.com/apache/spark/commit/8b18fa1c5a457198e3b99d41aaf770c8cb11106d).

SparkQA commented:
**[Test build #72230 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72230/testReport)** for PR 16758 at commit [`8b18fa1`](https://github.com/apache/spark/commit/8b18fa1c5a457198e3b99d41aaf770c8cb11106d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72230/
Test FAILed.

tdas committed:
Fixed bug

SparkQA commented:
**[Test build #72256 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72256/testReport)** for PR 16758 at commit [`8b3150a`](https://github.com/apache/spark/commit/8b3150a7bd74d70211ee0c8bcc993191b8cd25a3).

SparkQA commented:
**[Test build #72256 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72256/testReport)** for PR 16758 at commit [`8b3150a`](https://github.com/apache/spark/commit/8b3150a7bd74d70211ee0c8bcc993191b8cd25a3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72256/
Test PASSed.

lw-lin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/KeyedStateImpl.scala:
```diff
@@ -0,0 +1,57 @@
+    removed = true
+  }
+
+  override def toString: String = "KeyedState($value)"
```
nit: _s_"KeyedState($value)"

tdas replied:
done

lw-lin commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/MapGroupsWithStateSuite.scala:
```diff
@@ -0,0 +1,240 @@
+    // Function that returns running count only if its even, otherwise does not return
+    val stateFunc = (key: String, values: Iterator[String], state: KeyedState[RunningCount]) => {
+      if (state.exists) throw new IllegalArgumentException("state.exists should be false")
+      if (state.exists) {
```
nit: state.get.exists?

tdas replied:
done. not a nit! wrong test!

lw-lin commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/MapGroupsWithStateSuite.scala:
```diff
@@ -0,0 +1,240 @@
+  test("mapGroupsWithState - batch") {
+    val stateFunc = (key: String, values: Iterator[String], state: KeyedState[RunningCount]) => {
+      if (state.exists) throw new IllegalArgumentException("state.exists should be false")
+      if (state.exists) {
```
same here

tdas replied:
done. thanks for catching these.

lw-lin commented on sql/core/src/main/scala/org/apache/spark/sql/KeyedState.scala:
```diff
@@ -0,0 +1,134 @@
+ *  - If the `remove()` is called, then `exists()` will return `false`, and
+ *    `getOption()` will return `None`.
+ *  - After that `update(newState)` is called, then `exists()` will return `true`,
+ *    and `getOption()` will return `Some(...)`.
```
nit: getOption ...

zsxwing commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:
```diff
@@ -235,3 +234,79 @@ case class StateStoreSaveExec(
+        val groupedIter = GroupedIterator(iter, groupingAttributes, child.output)
+
+        val getKeyObj = ObjectOperator.deserializeRowToObject(keyDeserializer, groupingAttributes)
+        val getKey = GenerateUnsafeProjection.generate(groupingAttributes, child.output)
```
nit: not used

tdas replied:
done.

zsxwing commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:
```diff
@@ -235,3 +234,79 @@ case class StateStoreSaveExec(
+          val wrappedState = KeyedStateImpl[Any](stateObjOption.orNull)
+          val mappedIterator = func(keyObj, valueObjIter, wrappedState)
+
+          if (wrappedState.isRemoved) {
```
this should not be checked here. `mappedIterator` may be lazy. You can wrap it with a CompletionIterator and add these codes there.

tdas replied:
good point. this would confusing for the user if we dont do this.

zsxwing reviewed (changes requested):
Made one pass. Most of my comments are nits.

brkyvz commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala:
```diff
@@ -46,8 +46,13 @@ object UnsupportedOperationChecker {
 
+    /** Collect all the streaming aggregates in a sub plan */
+    def collectStreamingAggregates(subplan: LogicalPlan): Seq[Aggregate] = {
+      subplan.collect { case a@Aggregate(_, _, _) if a.isStreaming => a }
```
nit: space before and after a. Actually if you're not going to use any of the `Aggregate` params, just change this to `a: Aggregate`

tdas replied:
done.

brkyvz commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala:
```diff
@@ -313,6 +313,56 @@ case class MapGroups(
+    stateSerializer: Seq[NamedExpression],
+    child: LogicalPlan) extends UnaryNode with ObjectProducer
+
+
```
nit: extra line

tdas replied:
removed

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/KeyedState.scala:
```diff
@@ -0,0 +1,134 @@
+
+  /**
+   * Update the value of the state. Note that null is not a valid value, and `update(null)` is
+   * same as `remove()`
```
@marmbrus any thoughts on this?

marmbrus replied:
if get for a non-existent key returns `null` then I think its reasonable that `remove(key)` and `put(key, null)` are the same.

zsxwing replied:
@marmbrus The user may pass a null by mistake (e.g., just call some method that may return null but not be aware of it). It's pretty hard to debug such silent mistake.

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala:
```diff
@@ -68,7 +71,7 @@ class IncrementalExecution(
    * preparation walks the query plan.
    */
-  private var operatorId = 0
+  private val operatorId = new AtomicInteger(0)
```
not really. just a tiny bit cleaner to do `getAndIncrement` than using x and then calling `x += 1`

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala:
```diff
@@ -147,6 +147,23 @@ private[state] class HDFSBackedStateStoreProvider(
+          case Some(ValueRemoved(_, _)) =>
+          // Remove already in update map, no need to change
+        }
+      }
```
yeah. that's true. I should add tests for this StateStore operation.

brkyvz commented on sql/core/src/main/scala/org/apache/spark/sql/KeyedState.scala:
```diff
@@ -0,0 +1,134 @@
+ * Important points to note about using KeyedState.
+ *  - The value of the state cannot be null. So updating state with null is same as removing it.
+ *  - Operations on `KeyedState` are not thread-safe. This is to avoid memory barriers.
+ *  - If the `remove()` is called, then `exists()` will return `false`, and
```
nit: `If ``remove`` is called`. Remove `the`

tdas replied:
done.

brkyvz commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/KeyedStateImpl.scala:
```diff
@@ -0,0 +1,57 @@
+
+import org.apache.spark.sql.KeyedState
+
+/** Internal implementation of the [[KeyedState]] interface */
```
Would be nice to note here that this implementation is not thread safe

tdas replied:
i have mentioned that the trait KeyedState is not thread-safe

tdas commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:
```diff
@@ -111,6 +111,25 @@ class UnsupportedOperationsSuite extends SparkFunSuite {
 
+  // MapGroupsWithState: Not supported after a streaming aggregation
+  val att = new AttributeReference(name = "a", dataType = LongType)()
+  assertSupportedInStreamingPlan(
```
done

tdas commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/MapGroupsWithStateSuite.scala:
```diff
@@ -0,0 +1,240 @@
+    // Returns the data and the count if state is defined, otherwise does not return anything
+    val stateFunc = (key: String, values: Iterator[String], state: KeyedState[RunningCount]) => {
+
+      var count = Option(state.get).map(_.count).getOrElse(0L) + values.size
```
done

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/KeyedState.scala:
```diff
@@ -0,0 +1,134 @@
+ *  - The value of the state cannot be null. So updating state with null is same as removing it.
+ *  - Operations on `KeyedState` are not thread-safe. This is to avoid memory barriers.
+ *  - If the `remove()` is called, then `exists()` will return `false`, and
+ *    `getOption()` will return `None`.
```
done

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/KeyedState.scala:
```diff
@@ -0,0 +1,134 @@
+ *  - The key of the group.
+ *  - An iterator containing all the values for this key.
+ *  - A user-defined state object set by previous invocations of the given function.
+ * In case of a batch Dataset, there is only invocation and state object will be empty as
```
done

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/KeyedState.scala:
```diff
@@ -0,0 +1,134 @@
+ *    new MapGroupsWithStateFunction<String, Integer, Integer, String>() {
+ *
+ *      @Override
+ *      public String call(String key, Optional<Integer> value, KeyedState<Integer> state) {
```
done

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/KeyedState.scala:
```diff
@@ -0,0 +1,134 @@
+ * Scala example of using `KeyedState` in `mapGroupsWithState`:
+ * {{{
+ * // A mapping function that maintains an integer state for string keys and returns a string.
+ * def mappingFunction(key: String, value: Iterable[Int], state: KeyedState[Int]): Option[String]= {
```
done

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala:
```diff
@@ -58,6 +58,8 @@ trait StateStore {
    */
   def remove(condition: UnsafeRow => Boolean): Unit
 
+  def remove(key: UnsafeRow): Unit
```
done

tdas committed:
Addressed concerns

tdas committed:
Merge remote-tracking branch 'apache-github/master' into mapWithState

tdas commented:
I addressed all the comments. However, @zsxwing, our offline discussion of throwing error on `.update(null)` ran into a problem. Since its typed as S, the behavior is odd when S is primitive type. See the failing test. When the type is Int, `get` return 0 when the state does not exist. That's very non-intuitive.

SparkQA commented:
**[Test build #72490 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72490/testReport)** for PR 16758 at commit [`b32dcd1`](https://github.com/apache/spark/commit/b32dcd170a55c655c292d2cc535893c02a86c296).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72490/
Test FAILed.

tdas committed:
Addressed comments

SparkQA commented:
**[Test build #72543 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72543/testReport)** for PR 16758 at commit [`f3d1231`](https://github.com/apache/spark/commit/f3d12311229a958fee5fb3cfcf4f6b33035fc87f).

tdas commented:
@zsxwing had a discussion with @marmbrus. It is indeed weird that after calling `state.remove()` and `state.get` will return 0 if it is of type int (as in JVM null casted to int = 0). So we went back to the semantics of `get` throwing error when state does not exist.
This ensures the following semantics that @marmbrus originally desired.
- `update(null)` is not allowed - for primitive types, compile disallows it; for other types, it throw IllegalArgumentException
- Hence, `get` will never return `Some(null)`.

tdas commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/MapGroupsWithStateSuite.scala:
```diff
@@ -0,0 +1,240 @@
+      StartStream(),
+      CheckLastBatch(("a", 3L))     // task should not fail, and should show correct count
+    )
+  }
```
Added.

SparkQA commented:
**[Test build #72543 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72543/testReport)** for PR 16758 at commit [`f3d1231`](https://github.com/apache/spark/commit/f3d12311229a958fee5fb3cfcf4f6b33035fc87f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72543/
Test PASSed.

zsxwing commented:
LGTM. Merging to master and 2.1.

zsxwing commented:
It conflicts with 2.1. Could you submit a backport PR, please?
